Engelsystem / IllegalArgumentException: Unable to create converter for java.util.List<info.metadude.kotlin.library.engelsystem.models.Shift>

# Environment
- Build flavor: `ccc36c3`
- Since commit: ef53203d3e73e92f0947387e0b1e4253aca01726
- Branch: `master`

# How to reproduce
1. Launch a **release** build
2. Visit the _Settings_ screen
3. Enter a valid _Engelsystem_ URL
4. Return to the main screen

# Observed behavior
- Nothing happens, a warning is logged to the console.
```
java.lang.IllegalArgumentException: Unable to create converter for 
  java.util.List<info.metadude.kotlin.library.engelsystem.models.Shift>
    for method EngelsystemService.getShifts
    at retrofit2.Utils.methodError(Utils.java:52)
    at retrofit2.HttpServiceMethod.createResponseConverter(HttpServiceMethod.java:115)
    at retrofit2.HttpServiceMethod.parseAnnotations(HttpServiceMethod.java:82)
    at retrofit2.ServiceMethod.parseAnnotations(ServiceMethod.java:37)
    at retrofit2.Retrofit.loadServiceMethod(Retrofit.java:170)
    at retrofit2.Retrofit$1.invoke(Retrofit.java:149)
    at java.lang.reflect.Proxy.invoke(Proxy.java:1006)
    at $Proxy0.getShifts(Unknown Source)
    at info.metadude.android.eventfahrplan.engelsystem.EngelsystemNetworkRepository.load(EngelsystemNetworkRepository.kt:25)
    at nerd.tuxmobil.fahrplan.congress.repositories.AppRepository$loadShifts$1.invokeSuspend(AppRepository.kt:182)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

Caused by: java.lang.RuntimeException: Failed to find the generated JsonAdapter constructor
  for class info.metadude.kotlin.library.engelsystem.models.Shift
    at com.squareup.moshi.internal.Util.generatedAdapter(Util.java:591)
    at com.squareup.moshi.StandardJsonAdapters$1.create(StandardJsonAdapters.java:61)
    at com.squareup.moshi.Moshi.adapter(Moshi.java:145)
    at com.squareup.moshi.Moshi.adapter(Moshi.java:105)
    at com.squareup.moshi.Moshi.adapter(Moshi.java:74)
    at com.squareup.moshi.CollectionJsonAdapter.newArrayListAdapter(CollectionJsonAdapter.java:54)
    at com.squareup.moshi.CollectionJsonAdapter$1.create(CollectionJsonAdapter.java:38)
    at com.squareup.moshi.Moshi.adapter(Moshi.java:145)
    at com.squareup.moshi.Moshi.adapter(Moshi.java:105)
    at retrofit2.converter.moshi.MoshiConverterFactory.responseBodyConverter(MoshiConverterFactory.java:91)
    at retrofit2.Retrofit.nextResponseBodyConverter(Retrofit.java:330)
    at retrofit2.Retrofit.responseBodyConverter(Retrofit.java:313)
    at retrofit2.HttpServiceMethod.createResponseConverter(HttpServiceMethod.java:113)
    ... 14 more

Caused by: java.lang.NoSuchMethodException: info.metadude.kotlin.library.engelsystem.models.ShiftJsonAdapter.<init> []
    at java.lang.Class.getConstructor0(Class.java:2332)
    at java.lang.Class.getDeclaredConstructor(Class.java:2170)
    at com.squareup.moshi.internal.Util.generatedAdapter(Util.java:572)
    ... 26 more
```

# Expected behavior
- The _Engelsystem_ shifts show up.

# Quick workaround
- Downgrade Moshi from 1.11.0 to 1.9.3.

# Related
- https://github.com/square/moshi/issues/1266

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -45,6 +45,7 @@ android {
             minifyEnabled true
             proguardFiles getDefaultProguardFile('proguard-android.txt'),
                     'proguard-rules/proguard-project.txt',
+                    'proguard-rules/moshi.pro',
                     'proguard-rules/okhttp3.pro',
                     'proguard-rules/okio.pro'
         }

--- a/app/proguard-rules/moshi.pro
+++ b/app/proguard-rules/moshi.pro
@@ -1,0 +1,61 @@
+# JSR 305 annotations are for embedding nullability information.
+-dontwarn javax.annotation.**
+
+-keepclasseswithmembers class * {
+    @com.squareup.moshi.* <methods>;
+}
+
+-keep @com.squareup.moshi.JsonQualifier interface *
+
+# Enum field names are used by the integrated EnumJsonAdapter.
+# values() is synthesized by the Kotlin compiler and is used by EnumJsonAdapter indirectly
+# Annotate enums with @JsonClass(generateAdapter = false) to use them with Moshi.
+-keepclassmembers @com.squareup.moshi.JsonClass class * extends java.lang.Enum {
+    <fields>;
+    **[] values();
+}
+
+# The name of @JsonClass types is used to look up the generated adapter.
+-keepnames @com.squareup.moshi.JsonClass class *
+
+# Retain generated target class's synthetic defaults constructor and keep DefaultConstructorMarker's
+# name. We will look this up reflectively to invoke the type's constructor.
+#
+# We can't _just_ keep the defaults constructor because Proguard/R8's spec doesn't allow wildcard
+# matching preceding parameters.
+-keepnames class kotlin.jvm.internal.DefaultConstructorMarker
+-keepclassmembers @kotlin.Metadata @com.squareup.moshi.JsonClass class * {
+    synthetic <init>(...);
+}
+
+# Retain generated JsonAdapters if annotated type is retained.
+-if @com.squareup.moshi.JsonClass class *
+-keep class <1>JsonAdapter {
+    <init>(...);
+    <fields>;
+}
+-if @com.squareup.moshi.JsonClass class **$*
+-keep class <1>_<2>JsonAdapter {
+    <init>(...);
+    <fields>;
+}
+-if @com.squareup.moshi.JsonClass class **$*$*
+-keep class <1>_<2>_<3>JsonAdapter {
+    <init>(...);
+    <fields>;
+}
+-if @com.squareup.moshi.JsonClass class **$*$*$*
+-keep class <1>_<2>_<3>_<4>JsonAdapter {
+    <init>(...);
+    <fields>;
+}
+-if @com.squareup.moshi.JsonClass class **$*$*$*$*
+-keep class <1>_<2>_<3>_<4>_<5>JsonAdapter {
+    <init>(...);
+    <fields>;
+}
+-if @com.squareup.moshi.JsonClass class **$*$*$*$*$*
+-keep class <1>_<2>_<3>_<4>_<5>_<6>JsonAdapter {
+    <init>(...);
+    <fields>;
+}

--- a/engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/EngelsystemNetworkRepository.kt
+++ b/engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/EngelsystemNetworkRepository.kt
@@ -1,7 +1,6 @@
 package info.metadude.android.eventfahrplan.engelsystem
 
 import info.metadude.android.eventfahrplan.engelsystem.loading.ShiftsLoading.awaitShiftsResult
-import info.metadude.android.eventfahrplan.engelsystem.models.EngelsystemUri
 import info.metadude.android.eventfahrplan.engelsystem.models.ShiftsResult
 import info.metadude.android.eventfahrplan.engelsystem.utils.UriParser
 import info.metadude.kotlin.library.engelsystem.ApiModule
@@ -14,16 +13,15 @@ class EngelsystemNetworkRepository(
 
 ) {
 
-    suspend fun load(okHttpClient: OkHttpClient, url: String): ShiftsResult {
-        val uri: EngelsystemUri
-        try {
-            uri = uriParser.parseUri(url)
-        } catch (e: URISyntaxException) {
-            return ShiftsResult.Exception(e)
-        }
+    suspend fun load(okHttpClient: OkHttpClient, url: String) = try {
+        val uri = uriParser.parseUri(url)
         val service = ApiModule.provideEngelsystemService(uri.baseUrl, okHttpClient)
         val call = service.getShifts(uri.pathPart, uri.apiKey)
-        return call.awaitShiftsResult()
+        call.awaitShiftsResult()
+    } catch (e: URISyntaxException) {
+        ShiftsResult.Exception(e)
+    } catch (e: IllegalArgumentException) {
+        ShiftsResult.Exception(e)
     }
 
 }

